Add an ability to pass custom formatter function

iilyak opened the pull request.

iilyak commented:
Fix for https://github.com/ferd/recon/issues/23

aerosol commented on src/recon_trace.erl:
```diff
@@ -156,13 +156,16 @@
 
 -type matchspec()   :: [{[term()], [term()], [term()]}].
 -type shellfun()    :: fun((_) -> term()).
+-type formatterfun()    :: fun((_) -> string()).
```
align the neighbours? :)

aerosol replied:
also `-> iolist()` ?

iilyak commented:
Let me know when review is over so I can squash all commits before the merge.

ferd commented on src/recon_trace.erl:
```diff
@@ -207,6 +210,12 @@ calls({Mod, Fun, Args}, Max) ->
+%% @equiv calls(Spec, Tracer, Opts, fun recon_trace:format/1)
+-spec calls(tspec() | [tspec(),...], max(), options()) -> num_matches().
+calls(Spec, Tracer, Opts) ->
+    calls(Spec, Tracer, Opts, fun format/1).
```
How does that work? The type says `max()` for the second argument, but `Tracer` as a variable. I think you got some confusion in the API where `trace_call` is used to specify which process is tracing content.

ferd commented on src/recon_trace.erl:
```diff
@@ -156,32 +156,35 @@
--type max_rate()    :: {max_traces(), millisecs()}.
+-type matchspec()    :: [{[term()], [term()], [term()]}].
+-type shellfun()     :: fun((_) -> term()).
+-type formatterfun() :: fun((_) -> iolist()).
```
might be better to support `iodata()` so top-level binaries are supported.

ferd commented:
Additionally to the comments I have posted (I'm sorry they're late, I was on vacation), I would love to see some evidence showing that there is little to no difference in performance between running direct calls to `format/1` and an indirect one via `fun format/1`. This code runs in a tight loop that is critical to not crashing production nodes, and any non-essential slowdown that can be avoided should be avoided.

For a tool like recon, I'll favor a lack of flexibility to a lack of safety every day.

iilyak commented:
Here is the performance impact.

1) Modify source as follows
```
--- a/src/recon_trace.erl
+++ b/src/recon_trace.erl
@@ -359,7 +359,9 @@ formatter(Tracer, Leader, FormatterFun) ->
         {'EXIT', Tracer, Reason} ->
             exit(Reason);
         TraceMsg ->
-            io:format(Leader, FormatterFun(TraceMsg), []),
+            {T1, _} = timer:tc(FormatterFun, [TraceMsg]),
+            {T2, _} = timer:tc(fun() -> format(TraceMsg) end),
+            io:format(Leader, "~w, ~w~n", [T1, T2]),
             formatter(Tracer, Leader, FormatterFun)
     end.
```
2) start tracing
```
Tracefile = "recon-perf.csv".
{ok, F} = file:open(Tracefile, [write]).
group_leader(F, self()).
TraceSpec = [{'_', '_', fun(_) -> return_trace() end}].
recon_trace:calls(TraceSpec, 20000, [{pid, all}, return_trace]).
```
3) remove two first lines from Tracefile manualy before doing following commands
4) parse csv file
```
{ok, Bin} = file:read_file(Tracefile).
Lines = re:split(Bin, "\n", []).
Records = [re:split(L, ",", []) || L <- Lines].

Data1 = [list_to_integer(string:strip(binary_to_list(T))) || [T, _] <- Records].
Data2 = [list_to_integer(string:strip(binary_to_list(T))) || [_, T] <- Records].
```
5) analyze results
```
(node1@127.0.0.1)16> rp(bear:get_statistics(Data1)).
[{min,15},
 {max,1818},
 {arithmetic_mean,36.62613130656533},
 {geometric_mean,32.653279858179},
 {harmonic_mean,31.161203686223548},
 {median,29},
 {variance,2288.0491841069843},
 {standard_deviation,47.83355709234872},
 {skewness,18.386858892556223},
 {kurtosis,424.205948851075},
 {percentile,[{50,29},
              {75,39},
              {90,46},
              {95,54},
              {99,92},
              {999,878}]},
 {histogram,[{21,301},
             {27,6846},
             {33,5045},
             {39,3045},
             {45,2537},
             {55,1310},
             {65,518},
             {75,126},
             {85,49},
             {95,33},
             {105,8},
             {115,8},
             {125,11},
             {135,5},
             {145,6},
             {155,5},
             {165,3},
             {175,7},
             {185,2},
             {195,6},
             {205,1},
             {215,1},
             {225,6},
             {235,3},
             {245,6},
             {255,3},
             {265,2},
             {275,4},
             {285,4},
             {295,5},
             {305,6},
             {315,5},
             {325,6},
             {335,5},
             {415,19},
             {515,11},
             {615,4},
             {715,8},
             {815,7},
             {915,3},
             {1015,5},
             {1115,4},
             {1215,5},
             {1315,2},
             {1415,1},
             {1515,1},
             {1615,0},
             {1715,0},
             {1815,0},
             {1915,1}]},
 {n,19999}]
ok
(node1@127.0.0.1)17> rp(bear:get_statistics(Data2)).
[{min,13},
 {max,4380},
 {arithmetic_mean,30.425371268563428},
 {geometric_mean,25.90770767812466},
 {harmonic_mean,24.518775717719254},
 {median,22},
 {variance,4431.693487679308},
 {standard_deviation,66.57096580101049},
 {skewness,34.86735630178287},
 {kurtosis,1723.4796195590186},
 {percentile,[{50,22},
              {75,35},
              {90,40},
              {95,44},
              {99,74},
              {999,850}]},
 {histogram,[{22,11134},
             {31,3224},
             {40,4203},
             {53,1066},
             {63,136},
             {73,35},
             {83,27},
             {93,12},
             {103,5},
             {113,6},
             {123,8},
             {133,4},
             {143,7},
             {153,3},
             {163,3},
             {173,3},
             {183,4},
             {193,2},
             {203,1},
             {213,1},
             {223,3},
             {233,1},
             {243,2},
             {253,4},
             {263,2},
             {273,3},
             {283,3},
             {293,5},
             {303,6},
             {313,5},
             {323,1},
             {333,1},
             {413,24},
             {513,8},
             {613,5},
             {713,8},
             {813,7},
             {913,7},
             {1013,4},
             {1113,4},
             {1213,5},
             {1313,1},
             {1413,2},
             {1513,0},
             {1613,0},
             {1713,0},
             {1813,0},
             {1913,0},
             {2013,0},
             {2113,0},
             {2213,0},
             {2313,0},
             {2413,0},
             {2513,0},
             {2613,0},
             {2713,0},
             {2813,1},
             {2913,0},
             {3013,0},
             {3113,1},
             {3213,0},
             {4013,1},
             {5013,1}]},
 {n,19999}]
ok
```

iilyak commented:
I can do similar measurements using `timer:now_diff` instead of `timer:tc` if you like.

ferd commented:
That seems to be good in terms of perf. Some variation, but not enough to worry me yet.

ferd commented:
I'll merge this in, and whenever I have time to update the docs for that, I will, and eventually cut a new release.

iilyak commented:
I didn't have chance to squash.

iilyak commented:
Thank you for merging this in.

ferd commented:
Just a heads up if you were using `master` in production: I've reworked the patch in there so that the custom formatter is passed in as `{formatter, fun(_) -> io_data()}` instead of a standalone argument, so that it fit the existing mechanisms for options better.
